### PR TITLE
New fibbing features

### PR DESCRIPTION
**Pull request by lferran**
Rebased forked FibbingNode and cherry picked useful commits to have a cleaner history

The previous PR can be closed

### CODE REVIEW

File: fibbingnode/misc/mininetlib/iprouter.py [hunk 1]
```diff
@@ -98,6 +99,29 @@ def __init__(self, router, *args, **kwargs):
         self.ospf.redistribute.static = 1000
         self.ospf.router_id = router.id
 
+        # Parse LSA throttling parameters
+        delay = CFG.get("DEFAULT", 'delay')
+        delay = delay if delay != '' else 0
```
Comment (oliviertilmans) [inline fibbingnode/misc/mininetlib/iprouter.py]: The default value should be in the res/default.cfg if you intend to make this configurable, instead of handled here as a if then elsse

**Comment by oliviertilmans** [comment]
Thanks!
This looks much better!

File: fibbingnode/misc/mininetlib/iprouter.py (hunk 1 above)
Comment (lferran) [reply fibbingnode/misc/mininetlib/iprouter.py]: Allright :+1: I just wanted to make sure that the default values apply in case nothing is specified in default.cfg

Comment (oliviertilmans) [reply fibbingnode/misc/mininetlib/iprouter.py]: Putting the defaults in the res/default.cfg ensures that it exists, as it gets automatically overwritten in https://github.com/Fibbing/FibbingNode/blob/master/fibbingnode/southbound/main.py#L174 if the user specifies an alternate config

File: fibbingnode/misc/mininetlib/iprouter.py [hunk 2]
```diff
@@ -98,6 +99,22 @@ def __init__(self, router, *args, **kwargs):
...
+        # Parse LSA throttling parameters
+        delay = CFG.get("DEFAULT", 'delay')
+        initial_holdtime = CFG.get("DEFAULT", 'initial_holdtime')
+        max_holdtime = CFG.get("DEFAULT", 'max_holdtime')
+
+        # Parse minimum LS intervals
+        min_ls_interval = CFG.get("DEFAULT", 'min_ls_interval')
+        min_ls_arrival = CFG.get("DEFAULT", 'min_ls_arrival')
+
+        self.ospf.throttling = {'spf': {'delay': delay,
```
Comment (oliviertilmans) [inline fibbingnode/misc/mininetlib/iprouter.py]: You should build these using chains of ConfigDict() dictionary types, e.g.
self.ospf.throttling = ConfigDict(spf=ConfigDict(delay=delay, initial_holdtime=...),
lsa-all=ConfigDict() ...)
(not sure you can have a - in key names that way)

This will simplify the config templates.

File: fibbingnode/res/templates/ospf.mako [hunk 3]
```diff
@@ -35,4 +35,9 @@ router ospf
   % for itf in node.ospf.passive_interfaces:
   passive-interface ${itf.name}
   % endfor
+  % if node.ospf.throttling and node.ospf.lsa:
+  timers throttle spf ${node.ospf.throttling['spf'].get('delay')} ${node.ospf.throttling['spf'].get('initial_holdtime')} ${node.ospf.throttling['spf'].get('max_holdtime')}
```
Comment (oliviertilmans) [inline fibbingnode/res/templates/ospf.mako]: Why do you need those get() calls ? especially since you don't provide the default value if the key is not found as the config line will then be 'timers throttle spf None ...'

File: fibbingnode/res/templates/ospf.mako [hunk 4]
```diff
@@ -35,4 +35,9 @@ router ospf
...
+  timers throttle lsa all ${node.ospf.throttling['lsa-all'].get('min_ls_interval')}
```
Comment (oliviertilmans) [inline fibbingnode/res/templates/ospf.mako]: If you use the ConfigDict() as requested above, you can then simply this to
node.ospf.throttling.spf.delay

File: fibbingnode/southbound/fibbing.py [hunk 5]
```diff
@@ -323,6 +323,14 @@ def proxy_remove(self, points):
     def proxy_connected(self, session):
         self.root.send_lsdblog_to(session)
 
+        if self.routes:
+            log.info("Removing old routes from previous sessions...")
+            for r in self.routes.iterkeys():
+                try:
+                    self.remove_route(r)
+                except:
```
Comment (oliviertilmans) [inline fibbingnode/southbound/fibbing.py]: What kind of exception can happen there ?
It would be great if the exception was handled in remove_route directly so you could simplify this whole block as (including the if part):
map(self.remove_route, self.routes.iterkeys())

I suspect that one of the errors would be that the dict size has changed so you would need to change iterkeys() to keys()

File: fibbingnode/southbound/lsdb.py [hunk 6]
```diff
@@ -132,21 +133,29 @@ def endpoints(self, lsdb):
...
         self.mask = mask
         self.age = int(age)
+        self.lsa_seqnum = self.int_to_2comp(int(lsa_seqnum), 32)
+
+    @staticmethod
+    def int_to_2comp(value, bits):
+        """Compute the 2's compliment of int value val"""
+        if (value & (1 << (bits - 1))) != 0:
+            value = value - (1 << bits)
+        return value
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: You can simplify this by directly computing the 2-complement  and returning it (as you compute half of it anyway for the if):
mask = 2 ** (bits - 1)
return -(value & mask) + (value & ~mask)

**Review by oliviertilmans — CHANGES_REQUESTED** [review]

**Comment by oliviertilmans** [comment]
We're almost there!

Thanks for your work!

**Comment by lferran** [comment]
Regarding the removal of routes in fibbing.py, I tried to address what you proposed in the previous PR, but I didn't manage. You suggested:

1. Make removal of routes when the application disconnects
2. Remove only routes from the session that is about to disconnect

For the moment, though, it works good as well removing old routes in proxy_connected().

As I understand it:
Regarding 1: a new function proxy_disconnected() should be invoked in the SJMPServer when such event happens. I didn't suceed.

Regarding 2: Shouldn't we remove all routes anyway? Are there routes that are not given by the application?

**Comment by oliviertilmans** [comment]
I'll write that bit myself. The design of the controller enables to have multiples applications concurrently add/remove fake nodes, I'll rewrite those handlers.

File: fibbingnode/southbound/lsdb.py [hunk 7]
```diff
@@ -466,7 +505,30 @@ def process_lsa(self):
                                  for part in lsa_info.split(SEP_GROUP) if part]
                     hdr = LSAHeader.parse(lsa_parts.pop(0))
                     lsa = LSA.parse(hdr, lsa_parts)
+                    new_seqnum = hdr.lsa_seqnum
+
                     log.debug('Parsed %s: %s', action, lsa)
+
+                    # For now, we just log lsa seq_num and ignore an LSA when
+                    # it has an older seq_num than the one in the database
+                    if hdr.lsa_type == '5':
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: Could you expand the seqnum handling to all lsa types ? We basically want to keep our view of the LSDB totally in sync with the network, so applying the seqnum handling to all LSA types makes sense.

**Review by oliviertilmans — CHANGES_REQUESTED** [review]

File: fibbingnode/southbound/lsdb.py (hunk 7 above)
Comment (lferran) [reply fibbingnode/southbound/lsdb.py]: Sure :+1: 
Do you like the way the sequence number is currently saved? Do you think it is useful to have 3 different sequence number dictionaries (one for router, network and external lsa), or should we just have a single one?

Comment (lferran) [reply fibbingnode/southbound/lsdb.py]: When I log seq_num for all LSA types, and ignore those LSA received with an older seq_num than the one currently in the database... it seems that some of the R and N lsas arrive with an old seq_num, which 

`[  DEBUG] endpoints: Cannot resolve network lsa for 192.168.190.2 yet`
`[  DEBUG] process_lsa: Parsed REM: [N]<172.16.0.1: 172.16.0.1, 172.16.0.2>`
`[  DEBUG] process_lsa: Current seqnum for LSA in lsdb: -2147483641)`
`[  DEBUG] process_lsa: OLD seqnum for LSA: [N]<172.16.0.1: 172.16.0.1, 172.16.0.2>, ignoring it...`

Which makes these LSAs unresolvable:

`[  DEBUG] endpoints: Cannot resolve network lsa for 172.16.0.1 yet`
`[  DEBUG] endpoints: Cannot resolve network lsa for 172.16.0.1 yet`

File: fibbingnode/southbound/fibbing.py [hunk 8]
```diff
@@ -251,6 +251,9 @@ def remove_route(self, network):
         except KeyError:
             log.debug('No route for network %s', net)
 
+        except Exception as other_exception:
+            log.error("Other exception found: {0}".format(other_exception))
```
Comment (oliviertilmans) [inline fibbingnode/southbound/fibbing.py]: log.exception also exists and pretty print a stack trace.

**Review by oliviertilmans — CHANGES_REQUESTED** [review]

**Comment by oliviertilmans** [comment]
Looking more deeply at your LSA sequence number detection, you implemented the ordering for ospf v1, which used lollipop numbering.

OSPFv2 uses a linearly increasing range of values, starting at 0x80000001, as per https://tools.ietf.org/html/rfc2328#page-120

To detect newer LSAs, you thus need to interpret the seq_num as an unsigned number (which is done when extracting it from quagga), then check if it is < or > to the previous one. The LSA seq_num wrapping is performed when the seq_num reaches 0x7fffffff, where it is first flushed from the routing domain (age set to max age), then reinserted again with age set to 0x80000001.

**Comment by lferran** [comment]
Ok... my bad. So please, correct me if I am wrong, but from what I understood in the RFC:

- The seq_num value that is parsed from the quagga file should be interpreted as a signed integer, which means that I have to convert it using `int_to_comp(int(value), 32)`

- When deciding if seq_numA is newer than seq_numB, i should just do: `return seq_numA > seq_numB`

**Comment by oliviertilmans** [comment]
You don't need to do any conversion, as the number you get from Quagga is already correct (the rfc wants a signed 32 bit integer ranging from 0x80000000 to 0x7fffffff which is what is stored in lsa->ls_seqnum).

Comparing is thus simply a > b indeed (for a newer than b).
You just need to allow the lsa to be processed if it is at maxage and same seqnum than before (in order to be flushed from the db then added back). See comment on the diff.

File: fibbingnode/southbound/lsdb.py [hunk 9]
```diff
@@ -466,7 +503,29 @@ def process_lsa(self):
...
+                    c_seqnum = self.get_current_seq_number(lsa)
+                    if not c_seqnum:
+                        log.debug('LSA was not in the LSDB yet')
+                    else:
+                        log.debug('Seqnum for LSA in lsdb: %s)', c_seqnum)
+                        log.debug('Seqnum for newly arrived LSA: %s)', new_seqnum)
+                        if new_seqnum == c_seqnum:
+                            log.debug("DUPLICATE seqnum for new LSA: %s", lsa)
+
+                        elif self.is_newer_seqnum(new_seqnum, c_seqnum):
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: this in itself is simply:
elif new_seqnum > c_seqnum

File: fibbingnode/southbound/lsdb.py [hunk 10]
```diff
@@ -466,7 +503,29 @@ def process_lsa(self):
...
+                            log.debug("NEWER seqnum received")
+
+                        else:
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: This needs to become:
elif hdr.age < MAX_LS_AGE:

in order to support premature LSA aging to wrap the sequence number around

**Review by oliviertilmans — CHANGES_REQUESTED** [review]

File: fibbingnode/southbound/lsdb.py (hunk 10 above)
Comment (lferran) [inline fibbingnode/southbound/lsdb.py]: :+1:

File: fibbingnode/southbound/lsdb.py (hunk 9 above)
Comment (lferran) [reply fibbingnode/southbound/lsdb.py]: :+1:  However, I think we have to re-patch how the LSA seq_num is dumped in Quagga. If you look at ospfd/ospf_dump.c:L838:

`ospf_log_lsdb_write_field_unsigned(LOG_SEQNUM, ntohl(hdr->ls_seqnum));`

change it for:

 `ospf_log_lsdb_write_field (LOG_SEQNUM, ntohl(hdr->ls_seqnum));`

don't you think?

File: fibbingnode/southbound/lsdb.py [hunk 11]
```diff
@@ -521,6 +509,9 @@ def process_lsa(self):
                         elif self.is_newer_seqnum(new_seqnum, c_seqnum):
                             log.debug("NEWER seqnum received")
 
+                        elif hdr.age < MAX_LS_AGE:
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: If you put it this way then it becomes 

```python
elif new_seqnum == c_seqnum and hdr.age >= MAX_LS_AGE:
```

Which is also fine by me

**Review by oliviertilmans — CHANGES_REQUESTED** [review]

File: fibbingnode/southbound/lsdb.py (hunk 9 above)
Comment (oliviertilmans) [reply fibbingnode/southbound/lsdb.py]: 👍 spot on, see https://github.com/Fibbing/Quagga/commit/0127eddbfe8decf92d28f8ceff8f92ea79bcdf23

File: fibbingnode/southbound/lsdb.py [hunk 12]
```diff
@@ -466,16 +491,43 @@ def process_lsa(self):
...
+                    else:
+                        log.debug('Seqnum for LSA in lsdb: %s)', c_seqnum)
+                        log.debug('Seqnum for newly arrived LSA: %s)', new_seqnum)
+                        if new_seqnum == c_seqnum:
+                            log.debug("DUPLICATE seqnum for new LSA: %s", lsa)
+
+                        elif self.is_newer_seqnum(new_seqnum, c_seqnum):
+                            log.debug("NEWER seqnum received")
+
+                        elif hdr.age < MAX_LS_AGE:
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: This needs to change to
```python
new_seqnum == c_seqnum and hdr.age >= MAX_LS_AGE:
```

As you want to detect the wrap-around case (same seqnum and ls max age)

File: fibbingnode/southbound/lsdb.py [hunk 13]
```diff
@@ -466,16 +491,43 @@ def process_lsa(self):
...
+                            log.debug("Wrapping seqnum around for LSA: %s", lsa)
+
+                        else:
+                            log.debug("OLD seqnum for LSA: %s, ignoring it...", lsa)
+                            self.queue.task_done()
+                            continue
+
+                    if new_seqnum == c_seqnum and hdr.age >= MAX_LS_AGE:
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: Revert this to 

```python
if hdr.age >= MAX_LS_AGE:
```

As you want to flush max age LSA from the DB (either due to a wrap around, or simply because the routes are gone)

**Review by oliviertilmans — CHANGES_REQUESTED** [review]

File: fibbingnode/southbound/lsdb.py [hunk 14]
```diff
@@ -466,7 +491,38 @@ def process_lsa(self):
...
+                        log.debug('Seqnum for newly arrived LSA: %s)', new_seqnum)
+                        if new_seqnum == c_seqnum:
+                            log.debug("DUPLICATE seqnum for new LSA: %s", lsa)
+
+                            if hdr.age >= MAX_LS_AGE:
+                                log.debug("Wrapping seqnum around for LSA: %s", lsa)
+
+                            else:
+                                # What about in this case?
+                                pass
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: This whole if  log(duplicate) if  (wrap around) else ? Could become:
```python
if new_seqnum == c_seqnum:
    if hdr.age >= MAX_LS_AGE:
        log.debug("Premature aging to wrap seqnum around for LSA: %s", lsa)
    else:
        log.debug("DUPLICATE seqnum for new LSA: %s", lsa)
```

File: fibbingnode/southbound/fibbing.py [hunk 15]
```diff
@@ -251,6 +251,9 @@ def remove_route(self, network):
         except KeyError:
             log.debug('No route for network %s', net)
 
+        except Exception as other_exception:
+            log.exception(e)
```
Comment (oliviertilmans) [inline fibbingnode/southbound/fibbing.py]: change e to other_exception

**Review by oliviertilmans — CHANGES_REQUESTED** [review]
This should be the final changeset

File: fibbingnode/southbound/lsdb.py (hunk 14 above)
Comment (lferran) [reply fibbingnode/southbound/lsdb.py]: Thanks. I got a bit confused in the end... ;) 
However, don't you get some OLD LSA at network startup, and then the following logs?

`[  DEBUG] endpoints: Cannot resolve network lsa for 192.168.215.2 yet`
`[  DEBUG] endpoints: Cannot resolve network lsa for 192.168.221.2 yet`
`[  DEBUG] endpoints: Cannot resolve network lsa for 192.168.237.2 yet`
`[  DEBUG] endpoints: Cannot resolve network lsa for 192.168.238.2 yet`

Comment (oliviertilmans) [reply fibbingnode/southbound/lsdb.py]: Not sure I fully understand your issue there. Could paste a complete log somewhere ? Does the graph eventually gets build once the network has converged ?

**Comment by lferran** [comment]
Here is some parts of the (huge) complete log.

it seems that some Router LSA are received as OLD and then the network 
LSA cannot be resolved :/

I thing the graph gets built, because I can connect with my application 
and the HAS_INITIAL_GRAPH event gets set. However, I keep having these 
logs for ever:

...
[  DEBUG] endpoints: Cannot resolve network lsa for 192.168.255.2 yet
[  DEBUG] endpoints: Cannot resolve network lsa for 192.168.255.2 yet
[  DEBUG] endpoints: Cannot resolve network lsa for 192.168.31.2 yet
[  DEBUG] endpoints: Cannot resolve network lsa for 192.168.47.2 yet
[  DEBUG] endpoints: Cannot resolve network lsa for 172.16.0.1 yet
...

**Comment by oliviertilmans** [comment]
Could you upload that log somewhere ? I'd like to track down the
corresponding router lsa and see when they get ignored (i.e. too old, or
bad seqnum, or ...).

**Comment by lferran** [comment]
Here you go! 
[complete_log.txt](https://github.com/Fibbing/FibbingNode/files/611694/complete_log.txt)

**Comment by oliviertilmans** [comment]
The bug is caused by the fact that network LSAs are missing from the lsdb.
(sudo python -m fibbingnode ... wait 10s ... show_lsdb)

The issue is the way you keep track of the LSA seq nums.
You key your seqnum_db using the lsa.key() method. However that method enables to distinguish between lsa's of the _same type_. In this case, network LSAs and router LSAs share similar keys (the router id). Thus whenever you add a Network LSA in the LSDB, it gets discarded as there already is a router LSAs with a newer seqnum.

I'd suggest the following set of fixes (and will comment on the code snippets as well):
- Get rid of the global dict
- Embed the information in the LSA themselves
- When receiving a new LSA, try to get it from the lsdb.

Looking at the overall code, and to enforce seqnum ordering, I'd get rid of the whole transaction helper, as it just delays the actual installation (and never aborts).

File: fibbingnode/southbound/lsdb.py [hunk 16]
```diff
@@ -466,7 +491,35 @@ def process_lsa(self):
                                  for part in lsa_info.split(SEP_GROUP) if part]
                     hdr = LSAHeader.parse(lsa_parts.pop(0))
                     lsa = LSA.parse(hdr, lsa_parts)
+                    new_seqnum = hdr.lsa_seqnum
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py:487]: Directly put the seqnum in the LSA -- this will need a cleaner refactoring eventually but the current PR needs to be finished first.

lsa.seqnum = hdr.lsa_seqnum

File: fibbingnode/southbound/lsdb.py [hunk 17]
```diff
@@ -344,6 +346,7 @@ def __init__(self):
                                         name="lsa_processing_thread")
         self.processing_thread.setDaemon(True)
         self.processing_thread.start()
+        self.seqnums_db = {}
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: You won't need this

File: fibbingnode/southbound/lsdb.py [hunk 18]
```diff
@@ -430,20 +433,42 @@ def forwarding_address_of(self, src, dst):
 
     def remove_lsa(self, lsa):
         lsdb = self.lsdb(lsa)
+        # Update seq num db too
+        try:
+            del self.seqnums_db[lsa.key()]
+        except KeyError:
+            pass
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: Neither these 5 lines

File: fibbingnode/southbound/lsdb.py [hunk 19]
```diff
@@ -430,20 +433,42 @@ def forwarding_address_of(self, src, dst):
...
         except KeyError:
             pass
         except TypeError:
             pass  # LSDB is None
 
-    def add_lsa(self, lsa):
+    def add_lsa(self, lsa, seq_num):
         lsdb = self.lsdb(lsa)
+        # Update seq num
+        self.seqnums_db[lsa.key()] = seq_num
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: Same than before, revert the changes

File: fibbingnode/southbound/lsdb.py [hunk 20]
```diff
@@ -430,20 +433,42 @@ def forwarding_address_of(self, src, dst):
...
         try:
             lsdb[lsa.key()] = lsa
         except TypeError:
             pass  # LSDB is None
 
+    def get_current_seq_number(self, lsa):
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py:449]: Rewrite this function as:
```python
try:
    return self.lsdb(lsa)[lsa.key()].seqnum
except (KeyError, AttributeError):
    return None
```

File: fibbingnode/southbound/lsdb.py [hunk 21]
```diff
@@ -466,7 +491,35 @@ def process_lsa(self):
...
+
                     log.debug('Parsed %s: %s', action, lsa)
+
+                    # For now, we just log lsa seq_num and ignore an LSA when
+                    # it has an older seq_num than the one in the database
+                    c_seqnum = self.get_current_seq_number(lsa)
+                    if not c_seqnum:
+                        # If LSA wasn't in the lsdb: process it normally
+                        log.debug('LSA was not in the LSDB yet')
+
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: Let's try to reduce the log volume here, change the if not else to:

```python
if c_seqnum:
    [log.debug(Seqnum for LSA in lsdb]
```

File: fibbingnode/southbound/lsdb.py [hunk 22]
```diff
@@ -466,7 +491,35 @@ def process_lsa(self):
...
+                    else:
+                        log.debug('Seqnum for LSA in lsdb: %s)', c_seqnum)
+                        log.debug('Seqnum for newly arrived LSA: %s)', new_seqnum)
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: Reduce the log volume, we don't need these two log lines

File: fibbingnode/southbound/lsdb.py [hunk 23]
```diff
@@ -466,7 +491,35 @@ def process_lsa(self):
                                  for part in lsa_info.split(SEP_GROUP) if part]
                     hdr = LSAHeader.parse(lsa_parts.pop(0))
                     lsa = LSA.parse(hdr, lsa_parts)
+                    new_seqnum = hdr.lsa_seqnum
+
                     log.debug('Parsed %s: %s', action, lsa)
+
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py:490]: You can expand on the Parsed line:

```python
log.debug('Parsed %s: %s [%d]', action, lsa, lsa.seqnum)
```

File: fibbingnode/southbound/lsdb.py [hunk 24]
```diff
@@ -466,7 +491,35 @@ def process_lsa(self):
...
+                            if hdr.age >= MAX_LS_AGE:
+                                log.debug("Premature aging to wrap seqnum around for LSA: %s", lsa)
+                            else:
+                                log.debug("DUPLICATE seqnum for new LSA: %s", lsa)
+
+                        elif self.is_newer_seqnum(new_seqnum, c_seqnum):
+                            log.debug("NEWER seqnum received")
+
+                        else:
+                            log.debug("OLD seqnum for LSA: %s, ignoring it...", lsa)
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: Reduce the log volume (remove starting from: if new_seqnum == c_seqnum to here, then)

```python
if not self.is_newer_seqnum(new_seqnum, c_seqnum) and new_seqnum != c_seqnum:
   # Can get the same seqnum when flushing an LSA from the domain
    log.debug("OLD seqnum for LSA, ignoring it...")
    self.queue.task_done()
    continue
```

File: fibbingnode/southbound/lsdb.py [hunk 25]
```diff
@@ -475,7 +528,7 @@ def process_lsa(self):
                     if action == REM:
                         provider.remove_lsa(lsa)
                     elif action == ADD:
-                        provider.add_lsa(lsa)
+                        provider.add_lsa(lsa, new_seqnum)
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: You can revert the change.

File: fibbingnode/southbound/lsdb.py [hunk 26]
```diff
@@ -597,8 +651,8 @@ def commit(self, lsdb):
                       len(self.rem) + len(self.add))
         for lsa in self.rem:
             lsdb.remove_lsa(lsa)
-        for lsa in self.add:
-            lsdb.add_lsa(lsa)
+        for lsa, seqnum in self.add:
+            lsdb.add_lsa(lsa, seqnum)
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: You can revert

File: fibbingnode/southbound/lsdb.py [hunk 27]
```diff
@@ -585,8 +639,8 @@ def __init__(self):
         self.add = []
         self.rem = []
 
-    def add_lsa(self, lsa):
-        self.add.append(lsa)
+    def add_lsa(self, lsa, seqnum):
+        self.add.append((lsa, seqnum))
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: You can revert

**Review by oliviertilmans — CHANGES_REQUESTED** [review]
I'll remove the transaction helpers in another commit.
I highlighted the changes to solves the LSA key collision.

File: fibbingnode/southbound/lsdb.py (hunk 21 above)
Comment (lferran) [inline fibbingnode/southbound/lsdb.py]: I am not sure what you mean here.

File: fibbingnode/southbound/lsdb.py [hunk 28]
```diff
@@ -597,7 +629,7 @@ def commit(self, lsdb):
                       len(self.rem) + len(self.add))
         for lsa in self.rem:
             lsdb.remove_lsa(lsa)
-        for lsa in self.add:
+        for lsa, seqnum in self.add:
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: This will crash as you no longer have a seqnum outside of the lsa

File: fibbingnode/southbound/lsdb.py [hunk 29]
```diff
@@ -437,13 +439,28 @@ def remove_lsa(self, lsa):
         except TypeError:
             pass  # LSDB is None
 
-    def add_lsa(self, lsa):
+    def add_lsa(self, lsa, seq_num):
```
Comment (oliviertilmans) [inline fibbingnode/southbound/lsdb.py]: This will break as there aren't seqnum outside of lsa anymore

**Review by oliviertilmans — CHANGES_REQUESTED** [review]
There are 2 minor bugs due to the latest rewrite

**Review by oliviertilmans — APPROVED** [review]

**Comment by oliviertilmans** [comment]
Can you assess that thisfinal set of changes solve your earlier issue ? (the endpoints one). It did here.

**Comment by lferran** [comment]
Yes, it does also here :+1

**Comment by oliviertilmans** [comment]
Thanks!
